### PR TITLE
Added fallback to active python env when conda is absent

### LINKED ISSUE #337: Conda dependency on non-conda environment -- FileNotFoundError: [Errno 2] No such file or directory: 'conda'
## Description

Using Debian Stable and pyenv to install Python. I can't install it. I've tried setting it up a VPS but also on a VM using the official image from [Debian](https://www.debian.org/CD/netinst/) (amd64).

## Reproduce

Install Debian net install (https://www.debian.org/CD/netinst/), then
```
- su
- apt update
- apt full-upgrade
- # Documented python dependencies for pyenv and git, https://github.com/pyenv/pyenv/wiki
- apt update; apt install build-essential libssl-dev zlib1g-dev \
libbz2-dev libreadline-dev libsqlite3-dev curl \
libncursesw5-dev xz-utils tk-dev libxml2-dev libxmlsec1-dev libffi-dev liblzma-dev \
git
- exit
- git clone https://github.com/pyenv/pyenv.git ~/.pyenv
- cd ~/.pyenv && src/configure && make -C src
- cd
- echo 'export PYENV_ROOT="$HOME/.pyenv"' >> ~/.bashrc
- echo 'command -v pyenv >/dev/null || export PATH="$PYENV_ROOT/bin:$PATH"' >> ~/.bashrc
- echo 'eval "$(pyenv init -)"' >> ~/.bashrc
- echo 'export PYENV_ROOT="$HOME/.pyenv"' >> ~/.profile
- echo 'command -v pyenv >/dev/null || export PATH="$PYENV_ROOT/bin:$PATH"' >> ~/.profile
- echo 'eval "$(pyenv init -)"' >> ~/.profile
- exit
- <login again>
- pyenv install 3.11.1
- pyenv global 3.11.1
- python --version # confirming 3.11.1
- pip install --upgrade setuptools
- pip install jupyterlab jupyter_scheduler
- jupyter server --generate-config
- nano +328 .jupyter/jupyter_server_config.py 
- # modify `c.ServerApp.ip = 'localhost'`  to `c.ServerApp.ip = '*'`
- jupyter-lab
```

At this point it launches and you can create notebooks.

When you try to schedule it, I get errors,

Screen Shot 2023-01-24 at 1 27 17 PM

Screen Shot 2023-01-24 at 1 27 10 PM

**Most curious is** `FileNotFoundError: [Errno 2] No such file or directory: 'conda'`, on a non-conda environment.

```[E 2023-01-24 13:26:44.750 ServerApp] Uncaught exception GET /scheduler/runtime_environments?1674584804638 (192.168.138.125)
    HTTPServerRequest(protocol='http', host='192.168.138.46:8888', method='GET', uri='/scheduler/runtime_environments?1674584804638', version='HTTP/1.1', remote_ip='192.168.138.125')
    Traceback (most recent call last):
      File "/home/jupyter/.pyenv/versions/3.11.1/lib/python3.11/site-packages/tornado/web.py", line 1713, in _execute
        result = await result
                 ^^^^^^^^^^^^
      File "/home/jupyter/.pyenv/versions/3.11.1/lib/python3.11/site-packages/jupyter_scheduler/handlers.py", line 361, in get
        environments = await ensure_async(self.environments_manager.list_environments())
                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/jupyter/.pyenv/versions/3.11.1/lib/python3.11/site-packages/jupyter_scheduler/environments.py", line 35, in list_environments
        envs = subprocess.check_output(["conda", "env", "list", "--json"])
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/jupyter/.pyenv/versions/3.11.1/lib/python3.11/subprocess.py", line 466, in check_output
        return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/jupyter/.pyenv/versions/3.11.1/lib/python3.11/subprocess.py", line 548, in run
        with Popen(*popenargs, **kwargs) as process:
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/jupyter/.pyenv/versions/3.11.1/lib/python3.11/subprocess.py", line 1024, in __init__
        self._execute_child(args, executable, preexec_fn, close_fds,
      File "/home/jupyter/.pyenv/versions/3.11.1/lib/python3.11/subprocess.py", line 1901, in _execute_child
        raise child_exception_type(errno_num, err_msg, err_filename)
    FileNotFoundError: [Errno 2] No such file or directory: 'conda'
[W 2023-01-24 13:26:44.758 SchedulerApp] wrote error: 'Unhandled error'
    Traceback (most recent call last):
      File "/home/jupyter/.pyenv/versions/3.11.1/lib/python3.11/site-packages/tornado/web.py", line 1713, in _execute
        result = await result
                 ^^^^^^^^^^^^
      File "/home/jupyter/.pyenv/versions/3.11.1/lib/python3.11/site-packages/jupyter_scheduler/handlers.py", line 361, in get
        environments = await ensure_async(self.environments_manager.list_environments())
                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/jupyter/.pyenv/versions/3.11.1/lib/python3.11/site-packages/jupyter_scheduler/environments.py", line 35, in list_environments
        envs = subprocess.check_output(["conda", "env", "list", "--json"])
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/jupyter/.pyenv/versions/3.11.1/lib/python3.11/subprocess.py", line 466, in check_output
        return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/jupyter/.pyenv/versions/3.11.1/lib/python3.11/subprocess.py", line 548, in run
        with Popen(*popenargs, **kwargs) as process:
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/jupyter/.pyenv/versions/3.11.1/lib/python3.11/subprocess.py", line 1024, in __init__
        self._execute_child(args, executable, preexec_fn, close_fds,
      File "/home/jupyter/.pyenv/versions/3.11.1/lib/python3.11/subprocess.py", line 1901, in _execute_child
        raise child_exception_type(errno_num, err_msg, err_filename)
    FileNotFoundError: [Errno 2] No such file or directory: 'conda'
[E 2023-01-24 13:26:44.761 SchedulerApp] {
      "Host": "192.168.138.46:8888",
      "Accept": "*/*",
      "Referer": "http://192.168.138.46:8888/lab/tree/Untitled.ipynb",
      "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:109.0) Gecko/20100101 Firefox/109.0"
    }
[E 2023-01-24 13:26:44.761 SchedulerApp] 500 GET /scheduler/runtime_environments?1674584804638 (d7e7862e6c144b84bd93b089f709444d@192.168.138.125) 13.19ms referer=http://192.168.138.46:8888/lab/tree/Untitled.ipynb
```

## Expected behavior

I expect to be able to schedule a notebook.

## Context

- Operating System and version:  
```
jupyter@jupyter:~$ lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 11 (bullseye)
Release:        11
Codename:       bullseye
```
- Browser and version:
  - Chrome: 108.0.5359.124
  - Firefox: 109.0 (64-bit)
- Jupyter Server version:  
```
jupyter_client           7.4.9
jupyter_core             5.1.5
jupyter-events           0.6.3
jupyter_scheduler        1.2.0
jupyter_server           2.1.0
jupyter_server_terminals 0.4.4
jupyterlab               3.5.3
jupyterlab-pygments      0.2.2
jupyterlab_server        2.19.0
```

<details><summary>Troubleshoot Output</summary>
<pre>
$ jupyter troubleshoot
$PATH:
        /home/jupyter/.pyenv/versions/3.11.1/bin
        /home/jupyter/.pyenv/libexec
        /home/jupyter/.pyenv/plugins/python-build/bin
        /home/jupyter/.pyenv/shims
        /home/jupyter/.pyenv/bin
        /usr/local/bin
        /usr/bin
        /bin
        /usr/local/games
        /usr/games

sys.path:
        /home/jupyter/.pyenv/versions/3.11.1/bin
        /home/jupyter/.pyenv/versions/3.11.1/lib/python311.zip
        /home/jupyter/.pyenv/versions/3.11.1/lib/python3.11
        /home/jupyter/.pyenv/versions/3.11.1/lib/python3.11/lib-dynload
        /home/jupyter/.pyenv/versions/3.11.1/lib/python3.11/site-packages

sys.executable:
        /home/jupyter/.pyenv/versions/3.11.1/bin/python3.11

sys.version:
        3.11.1 (main, Jan 24 2023, 13:16:00) [GCC 10.2.1 20210110]

platform.platform():
        Linux-5.10.0-21-amd64-x86_64-with-glibc2.31

which -a jupyter:
        /home/jupyter/.pyenv/versions/3.11.1/bin/jupyter
        /home/jupyter/.pyenv/shims/jupyter

pip list:
        Package                  Version
        ------------------------ -----------
        aiobotocore              2.4.2
        aiohttp                  3.8.3
        aioitertools             0.11.0
        aiosignal                1.3.1
        anyio                    3.6.2
        argon2-cffi              21.3.0
        argon2-cffi-bindings     21.2.0
        arrow                    1.2.3
        asttokens                2.2.1
        async-timeout            4.0.2
        attrs                    22.2.0
        Babel                    2.11.0
        backcall                 0.2.0
        beautifulsoup4           4.11.1
        bleach                   6.0.0
        botocore                 1.27.59
        certifi                  2022.12.7
        cffi                     1.15.1
        charset-normalizer       2.1.1
        comm                     0.1.2
        croniter                 1.3.8
        debugpy                  1.6.6
        decorator                5.1.1
        defusedxml               0.7.1
        entrypoints              0.4
        executing                1.2.0
        fastjsonschema           2.16.2
        fqdn                     1.5.1
        frozenlist               1.3.3
        fsspec                   2023.1.0
        greenlet                 2.0.1
        idna                     3.4
        ipykernel                6.20.2
        ipython                  8.8.0
        ipython-genutils         0.2.0
        isoduration              20.11.0
        jedi                     0.18.2
        Jinja2                   3.1.2
        jmespath                 1.0.1
        json5                    0.9.11
        jsonpointer              2.3
        jsonschema               4.17.3
        jupyter_client           7.4.9
        jupyter_core             5.1.5
        jupyter-events           0.6.3
        jupyter_scheduler        1.2.0
        jupyter_server           2.1.0
        jupyter_server_terminals 0.4.4
        jupyterlab               3.5.3
        jupyterlab-pygments      0.2.2
        jupyterlab_server        2.19.0
        MarkupSafe               2.1.2
        matplotlib-inline        0.1.6
        mistune                  2.0.4
        multidict                6.0.4
        nbclassic                0.4.8
        nbclient                 0.7.2
        nbconvert                7.2.9
        nbformat                 5.7.3
        nest-asyncio             1.5.6
        notebook                 6.5.2
        notebook_shim            0.2.2
        packaging                23.0
        pandocfilters            1.5.0
        parso                    0.8.3
        pexpect                  4.8.0
        pickleshare              0.7.5
        pip                      22.3.1
        platformdirs             2.6.2
        prometheus-client        0.16.0
        prompt-toolkit           3.0.36
        psutil                   5.9.4
        ptyprocess               0.7.0
        pure-eval                0.2.2
        pycparser                2.21
        pydantic                 1.10.4
        Pygments                 2.14.0
        pyrsistent               0.19.3
        python-dateutil          2.8.2
        python-json-logger       2.0.4
        pytz                     2022.7.1
        PyYAML                   6.0
        pyzmq                    25.0.0
        requests                 2.28.2
        rfc3339-validator        0.1.4
        rfc3986-validator        0.1.1
        s3fs                     2023.1.0
        Send2Trash               1.8.0
        setuptools               66.1.1
        six                      1.16.0
        sniffio                  1.3.0
        soupsieve                2.3.2.post1
        SQLAlchemy               1.4.46
        stack-data               0.6.2
        terminado                0.17.1
        tinycss2                 1.2.1
        tomli                    2.0.1
        tornado                  6.2
        traitlets                5.8.1
        typing_extensions        4.4.0
        uri-template             1.2.0
        urllib3                  1.26.14
        wcwidth                  0.2.6
        webcolors                1.12
        webencodings             0.5.1
        websocket-client         1.4.2
        wrapt                    1.14.1
        yarl                     1.8.2
</pre>
</details>

<details><summary>Command Line Output</summary>
<pre>
Paste the output from your command line running `jupyter lab` here, use `--debug` if possible.

Body too long, https://gist.github.com/rubenvarela/13c8302e218da2b3590679611c6bb4da#file-jupyter-lab-debug-md
</pre>
</details>

<details><summary>Browser Output</summary>

<pre>
Paste the output from your browser Javascript console here, if applicable.

Body too long, https://gist.github.com/rubenvarela/13c8302e218da2b3590679611c6bb4da#file-browser-console-md

Console output, 

Screen Shot 2023-01-24 at 2 20 22 PM

Network HTTP 500

Screen Shot 2023-01-24 at 2 20 50 PM

Network HTTP 500 Response
Screen Shot 2023-01-24 at 2 21 07 PM

Network HTTP 500 Stacktrace
Screen Shot 2023-01-24 at 2 21 23 PM
</pre>

</details>

### PR DESCRIPTION
**Pull request by 3coins**
Fixes #337

### CODE REVIEW

File: jupyter_scheduler/environments.py [hunk 1]
```diff
@@ -26,7 +27,10 @@ def output_formats_mapping(self) -> Dict[str, str]:
 
 
 class CondaEnvironmentManager(EnvironmentManager):
-    """Provides list of system installed conda environments"""
+    """Provides list of system installed conda environments,
+    falls back to the python executable that jupyter lab is
+    running in, if conda is not installed or activated
+    """
```
Comment (dlqqq) [inline jupyter_scheduler/environments.py:33]: Would suggest the following change to the docstring:

> Provides a list of Conda environments. If Conda is not installed or activated, it defaults to providing exclusively the Python executable that JupyterLab is currently running in.

**Review by dlqqq — CHANGES_REQUESTED** [review]
I think this change is safe; I don't see why a CondaEnvironmentManager needs to fail in the absence of Conda. However, I would like others to confirm this change works before merging.

File: jupyter_scheduler/environments.py (hunk 1 above)
Comment (3coins) [reply jupyter_scheduler/environments.py:33]: Thanks, that sounds much better.

**Comment by rubenvarela** [comment]
This is close.

On the same environment as defined here, https://github.com/jupyter-server/jupyter-scheduler/issues/337

Running this with `jupyter-lab --debug`,

```
[E 2023-02-24 14:54:01.684 ServerApp] Uncaught exception GET /scheduler/runtime_environments?1677268441540 (192.168.138.125)
    HTTPServerRequest(protocol='http', host='192.168.138.46:8888', method='GET', uri='/scheduler/runtime_environments?167726844154
0', version='HTTP/1.1', remote_ip='192.168.138.125')
    Traceback (most recent call last):
      File "/home/jupyter/.pyenv/versions/3.11.1/lib/python3.11/site-packages/tornado/web.py", line 1713, in _execute
        result = await result
                 ^^^^^^^^^^^^
      File "/home/jupyter/.pyenv/versions/3.11.1/lib/python3.11/site-packages/jupyter_scheduler/handlers.py", line 361, in get
        environments = await ensure_async(self.environments_manager.list_environments())
                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/jupyter/.pyenv/versions/3.11.1/lib/python3.11/site-packages/jupyter_scheduler/environments.py", line 36, in list
_environments
        envs = subprocess.check_output(["conda", "env", "list", "--json"])
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/jupyter/.pyenv/versions/3.11.1/lib/python3.11/subprocess.py", line 466, in check_output
        return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/jupyter/.pyenv/versions/3.11.1/lib/python3.11/subprocess.py", line 548, in run
        with Popen(*popenargs, **kwargs) as process:
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/jupyter/.pyenv/versions/3.11.1/lib/python3.11/subprocess.py", line 1024, in __init__
        self._execute_child(args, executable, preexec_fn, close_fds,
      File "/home/jupyter/.pyenv/versions/3.11.1/lib/python3.11/subprocess.py", line 1901, in _execute_child
        raise child_exception_type(errno_num, err_msg, err_filename)
    FileNotFoundError: [Errno 2] No such file or directory: 'conda'
```

Running the code in a notebook within the same environment, 
Screenshot 2023-02-24 at 2 56 18 PM

and the bottom portion,
Screenshot 2023-02-24 at 2 59 34 PM


----

Having said that, changing the exception to `FileNotFoundError` works and catching `stderr` works, 

```python
try:
    envs = subprocess.check_output(["conda", "env", "list", "--json"], stderr= subprocess.STDOUT)
    envs = json.loads(envs).get("envs", [])
except FileNotFoundError as e:
    print('caught')
```

Example,

Screenshot 2023-02-24 at 3 09 05 PM

**Comment by rubenvarela** [comment]
Doing the changes to the package locally, I can schedule a notebook. Running it every minute, it runs and the output works.

**Comment by 3coins** [comment]
@rubenvarela 
Thanks for suggesting that change. I have added another catch block for `FileNotFoundError`. Can you pull out the latest changes and re-test. Thanks for your help!

**Comment by rubenvarela** [comment]
@3coins it works!

**Review by dlqqq — APPROVED** [review]
Looks great, awesome work. @rubenvarela Thank you for testing this locally!

**Pull request merged**